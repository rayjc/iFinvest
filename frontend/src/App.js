import React, { useEffect } from 'react';
import logo from './logo.svg';
import './App.css';
import { authenticate } from './helpers/auth';
import { useDispatch } from 'react-redux';
import { loadUser } from './reducers/user/actions';

function App() {
  const dispatch = useDispatch();

  useEffect(() => {
    // check if user is logged in
    const token = authenticate();
    if (token) {
      // fetch and store user info to store
      dispatch(loadUser(token));
    }
  }, [dispatch]);

  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        <a
          className="App-link"
          href="https://reactjs.org"
          target="_blank"
          rel="noopener noreferrer"
        >
          Learn React
        </a>
      </header>
    </div>
  );
}

export default App;
