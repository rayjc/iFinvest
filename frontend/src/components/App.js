import React, { useEffect } from 'react';
import { authenticate } from '../helpers/auth';
import { useDispatch, useSelector } from 'react-redux';
import { loadUser } from '../reducers/user/actions';
import { loadPortfolios } from '../reducers/portfolios/actions';
import { ThemeProvider } from '@material-ui/core/styles';
import { createMuiTheme } from '@material-ui/core/styles';
import { blue, pink } from '@material-ui/core/colors';
import Menu from './Menu';
import Routes from './Routes';


const theme = createMuiTheme({
  palette: {
    primary: blue,
    secondary: pink,
  },
});

function App() {
  const dispatch = useDispatch();
  const user = useSelector(state => state.user.user);

  useEffect(() => {
    // check if user is logged in
    const token = authenticate();
    if (token) {
      // fetch and store user info to store
      dispatch(loadUser(token));
    }
  }, [dispatch]);

  useEffect(() => {
    if (user) {
      dispatch(loadPortfolios());
    }
  }, [user, dispatch]);

  return (
    <ThemeProvider theme={theme}>
      <Menu />
      <main>
        <Routes />
      </main>
    </ThemeProvider>
  );
}

export default App;
