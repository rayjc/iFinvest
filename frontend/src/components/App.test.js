import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render } from '@testing-library/react';
import App from './App';
import { useSelector, useDispatch } from 'react-redux';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch
}));

beforeEach(function() {
  useSelector.mockImplementation((selectorFn) =>
    selectorFn({
      user: {
        user: {
          id: 0, username: "test", first_name: "test", last_name: "user",
          email: "test@test.com"
        },
        error: null, isFetching: false,
      },
      portfolios: {
        portfolios: [
          { id: 0, name: "alpha", user_id: 0, created_at: "2020-01-10", investments: [] },
        ],
        error: null, isFetching: false,
      },
      investments: {
        investments: {}, error: null, isFetching: false,
      }
    }
    )
  );
});


// smoke test
it("renders without crashing", function() {
  render(<MemoryRouter><App /></MemoryRouter>);
});

// snapshot test
it("matches snapshot", function() {
  const { asFragment } = render(<MemoryRouter><App /></MemoryRouter>);
  expect(asFragment()).toMatchSnapshot();
});



describe("NavBar links test", function() {
  it("goes to /profile", async function() {
    const { getByText, findByLabelText } = render(<MemoryRouter><App /></MemoryRouter>);
    const profileLink = await findByLabelText("account of current user");

    fireEvent.click(profileLink);
    expect(getByText("First Name")).toBeInTheDocument();
    expect(getByText("Last Name")).toBeInTheDocument();
    expect(getByText("Email Address")).toBeInTheDocument();
  });
});

describe("Drawer links test", function() {
  it("goes to /profile", async function() {
    const { getByText, findByLabelText } = render(<MemoryRouter><App /></MemoryRouter>);
    const menuButton = await findByLabelText("menu");
    fireEvent.click(menuButton);
    const profileLink = getByText("Profile");
    fireEvent.click(profileLink);

    expect(getByText("First Name")).toBeInTheDocument();
    expect(getByText("Last Name")).toBeInTheDocument();
    expect(getByText("Email Address")).toBeInTheDocument();
  });

  it("goes to /:portfolioId", async function() {
    const { getAllByText, getByText, findByLabelText } = render(<MemoryRouter><App /></MemoryRouter>);
    const menuButton = await findByLabelText("menu");
    fireEvent.click(menuButton);
    const portfolioLinks = getAllByText("alpha");
    fireEvent.click(portfolioLinks[0]);

    expect(getByText("No investments yet!")).toBeInTheDocument();
  });
});