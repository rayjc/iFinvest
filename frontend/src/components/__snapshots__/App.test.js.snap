// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="makeStyles-root-16"
>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        aria-label="menu"
        class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-17 MuiIconButton-colorInherit MuiIconButton-edgeStart"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h6
        class="MuiTypography-root makeStyles-title-18 MuiTypography-h6"
      >
        iFinvest
      </h6>
      <a
        aria-disabled="false"
        aria-label="account of current user"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        href="/profile"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
      <button
        aria-label="log out"
        class="MuiButtonBase-root MuiIconButton-root makeStyles-logoutButton-19 MuiIconButton-colorInherit"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;
